File: src/main/java/bad/robot/radiate/teamcity/TeamCity.java
package bad.robot.radiate.teamcity;

import bad.robot.http.Headers;
import bad.robot.http.HttpClient;
import bad.robot.http.HttpResponse;
import bad.robot.radiate.Unmarshaller;
import com.googlecode.totallylazy.Callable1;
import com.googlecode.totallylazy.Either;
import com.googlecode.totallylazy.Sequence;

import java.net.URL;

import static bad.robot.http.HeaderList.headers;
import static bad.robot.http.HeaderPair.header;
import static bad.robot.radiate.teamcity.BuildLocatorBuilder.latest;
import static bad.robot.radiate.teamcity.BuildLocatorBuilder.running;
import static bad.robot.radiate.teamcity.TeamCityEndpoints.*;
import static com.googlecode.totallylazy.Monad.methods.sequenceE;
import static com.googlecode.totallylazy.Sequences.flatten;
import static com.googlecode.totallylazy.Sequences.sequence;

class TeamCity {

    private final Headers headers = headers(header("Accept", "application/json"));
    private final Server server;
    private final HttpClient http;
    private final Authorisation authorisation;
    private final Unmarshaller<HttpResponse, Iterable<Project>> projects;
    private final Unmarshaller<HttpResponse, Project> project;
    private final Unmarshaller<HttpResponse, Build> build;

    public TeamCity(Server server, Authorisation authorisation, HttpClient http, Unmarshaller<HttpResponse, Iterable<Project>> projects, Unmarshaller<HttpResponse, Project> project, Unmarshaller<HttpResponse, Build> build) {
        this.server = server;
        this.http = http;
        this.projects = projects;
        this.project = project;
        this.build = build;
        this.authorisation = authorisation;
    }

    public Iterable<Project> retrieveProjects() {
        URL url = server.urlFor(projectsEndpointFor(authorisation));
        HttpResponse response = http.get(url, headers);
        if (response.ok())
            return projects.unmarshall(response);
        throw new UnexpectedResponse(url, response);
    }

    public Iterable<Project> retrieveFullProjects(Iterable<Project> projects) {
        Either<TeamCityException, Sequence<Project>> expanded = sequenceE(sequence(projects).mapConcurrently(expandingToFullProject()));
        if (expanded.isLeft())
            throw expanded.left();
        return expanded.right();
    }

    public Iterable<BuildType> retrieveBuildTypes(Iterable<Project> projects) {
        return flatten(retrieveFullProjects(projects));
    }

    public Build retrieveLatestBuild(BuildType buildType) {
        URL url = server.urlFor(running(buildType), authorisation);
        HttpResponse response = http.get(url, headers);
        if (response.getStatusCode() == 404)
            return retrieveBuild(latest(buildType));
        if (response.ok())
            return build.unmarshall(response);
        throw new UnexpectedResponse(url, response);
    }

    private Build retrieveBuild(BuildLocatorBuilder locator) {
        URL url = server.urlFor(locator, authorisation);
        HttpResponse response = http.get(url, headers);
        if (response.ok())
            return build.unmarshall(response);
        if (response.getStatusCode() == 404)
            return new NoBuild();
        throw new UnexpectedResponse(url, response);
    }

    private Callable1<Project, Either<TeamCityException, Project>> expandingToFullProject() {
        return new Callable1<Project, Either<TeamCityException, Project>>() {
            @Override
            public Either<TeamCityException, Project> call(Project project) throws Exception {
                URL url = server.urlFor(project);
                HttpResponse response = http.get(url, headers);
                if (response.ok())
                    return Either.right(TeamCity.this.project.unmarshall(response));
                ;
                return Either.<TeamCityException, Project>left(new UnexpectedResponse(url, response));
            }
        };
    }

    public static class Functions {
        public static Callable1<BuildType, Build> toBuild(final TeamCity teamcity) {
            return new Callable1<BuildType, Build>() {
                @Override
                public Build call(BuildType buildType) throws Exception {
                    return teamcity.retrieveLatestBuild(buildType);
                }
            };
        }
    }
}

File: src/main/scala/bad/robot/radiate/teamcity/TeamCity.scala
package bad.robot.radiate.teamcity

import java.net.URL

import bad.robot.http.HeaderList._
import bad.robot.http.HeaderPair.header
import bad.robot.http.{HttpClient, HttpResponse => SimpleHttpResponse}
import bad.robot.radiate.Sequence.sequence
import bad.robot.radiate.teamcity.BuildLocatorBuilder.{latest, running}
import bad.robot.radiate.teamcity.TeamCityEndpoints._
import bad.robot.radiate._
import bad.robot.radiate.Error
import bad.robot.radiate.teamcity.HttpClientSyntax._

import scalaz.{-\/, \/-, \/}
import scalaz.syntax.either._

object HttpClientSyntax {
  implicit class HttpClientOps(http: HttpClient) {
    implicit def get[A](url: URL, success: SimpleHttpResponse => Error \/ A): Error \/ A = {
      val response = http.get(url, headers(header("Accept", "application/json")))
      response match {
        case response@HttpResponse(200) => success(response)
        case _ => UnexpectedResponse(url, response).left
      }
    }
  }
}

object TeamCity {
  def apply(server: TeamCityUrl, authorisation: Authorisation, http: HttpClient, projects: Unmarshaller[SimpleHttpResponse, Iterable[Project]], project: Unmarshaller[SimpleHttpResponse, Project], build: Unmarshaller[SimpleHttpResponse, Build]) = {
    new TeamCity(server, authorisation, http, projects, project, build)
  }
}

class TeamCity(server: TeamCityUrl, authorisation: Authorisation, http: HttpClient, projects: Unmarshaller[SimpleHttpResponse, Iterable[Project]], project: Unmarshaller[SimpleHttpResponse, Project], build: Unmarshaller[SimpleHttpResponse, Build]) {

  private val asJson = headers(header("Accept", "application/json"))

  def retrieveProjects: Error \/ Iterable[Project] = {
    val url = server.urlFor(projectsEndpointFor(authorisation))
    http.get(url, projects.unmarshall _)
  }

  def retrieveFullProjects(projects: Iterable[Project]): Error \/ List[Project] = {
    sequence(projects.par.map(expandingToFullProject).toList).leftMap(AggregateError)
  }

  def retrieveBuildTypes(projects: Iterable[Project]): Error \/ List[BuildType] = {
    retrieveFullProjects(projects).map(_.flatMap(_.buildTypes.list))
  }

  def retrieveLatestBuild(buildType: BuildType): Error \/ Build = {
    val url = server.urlFor(running(buildType), authorisation)
    http.get(url, asJson) match {
      case HttpResponse(404) => retrieveBuild(latest(buildType))
      case response @ HttpResponse(200) => build.unmarshall(response)
      case response => throw new UnexpectedHttpResponse(url, response)
    }
  }

  private def retrieveBuild(locator: BuildLocatorBuilder): Error \/ Build = {
    val url = server.urlFor(locator, authorisation)
    http.get(url, asJson) match {
      case HttpResponse(404) => \/-(new NoBuild)
      case response @ HttpResponse(200) => build.unmarshall(response)
      case response => -\/(UnexpectedResponse(url, response))
    }
  }

  private def expandingToFullProject: Project => Error \/ Project = project => {
    http.get(server.urlFor(project), TeamCity.this.project.unmarshall _)
  }
}